### PR TITLE
C++: Factor must-flow predicates out of two queries

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR factors out the common functionality from `cpp/unsafe-use-of-this` and `cpp/return-stack-allocated-memory`. This shouldn't change any behavior[^1], but it should mean that we re-evaluate a _tiny_ little bit less stuff in the Code Scanning suite.

[^1]: Apart from the path explanation for `cpp/unsafe-use-of-this` which is now a lot better as it doesn't just use the `toString` given by the IR.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+private predicate flowsFromSource(DataFlow::Node node, MustFlowConfiguration config) {
+  config.isSource(node)
+  or
+  exists(DataFlow::Node mid |
+    step(mid, node, config) and
+    flowsFromSource(mid, pragma[only_bind_into](config))
+  )
+}
+
+/** Holds if `instr` flows to a sink. */
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll:64]:
```suggestion
/** Holds if `node` flows to a sink. */
```

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  ) {
+    exists(int n |
+      parameterOf(_, f, n) and
+      flowIntoParameter(f, pragma[only_bind_into](pragma[only_bind_out](n)), call, instr, init)
+    )
+  }
+
+  /**
+   * Holds if `instr` is the qualifier to a call to the non-virtual function `f`, and
+   * `init` is the corresponding initiazation instruction that receives the value of
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll:115]:
```suggestion
   * `init` is the corresponding initialization instruction that receives the value of
```

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql [hunk 3]
```diff
@@ -18,157 +18,70 @@ import cpp
...
-  or
-  operandToInstructionStep(node1.asOperand(), node2.asInstruction())
-}
-
-predicate instructionToOperandStep(Instruction instr, Operand operand) { operand.getDef() = instr }
-
-/**
- * Holds if `operand` flows to the result of `instr`.
- *
- * This predicate ignores flow through `PhiInstruction`s to create a 'must flow' relation. It also
```
Comment (jketema) [inline cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:69]: This remark about `PhiInstruction`s is not in the new version, maybe add it to `MustFlow.qll`?

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql [hunk 4]
```diff
@@ -18,157 +18,70 @@ import cpp
...
+  MustFlowPathNode source, MustFlowPathNode sink, VariableAddressInstruction var,
+  ReturnStackAllocatedMemoryConfig conf
 where
-  hasFlow(source, sink) and
-  source.getNode().asInstruction() = var
+  conf.hasFlowPath(source, sink) and
+  source.getNode().asInstruction() = var and
+  // Only raise an alert if we're returning from the _same_ callable as the on that
+  // declared the stack variable.
+  var.getEnclosingFunction() = sink.getNode().getEnclosingCallable()
```
Comment (jketema) [inline cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:85]: This seems new, or what this somewhere else before in a disguise that I didn't pick up on?

**Review by jketema — COMMENTED** [review]
Only a few small comments.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 3 above)
Comment (jketema) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:69]: It think it's important to have this, as this is what locally makes this "must" flow.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 4 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:85]: Yeah, it's new. The old must-flow predicates in `cpp/return-stack-allocated-memory` didn't have flow into function calls, but that was implemented (and indeed required) for `cpp/unsafe-use-of-this`. So I've ported flow into function calls to this new library.

This means that we now track flow into function calls in `cpp/return-stack-allocated-memory` as well, and we could now get a result from this snippet without this extra condition:
```
void id(int* px) {
  return p; // this returns the local variable `x`, but it's fine as the local variable isn't declared in this scope.
}
void f() {
  int x;
  int* px = id(&x);
}
```

Hm, I should actually add this test to the test cases for the `cpp/return-stack-allocated-memory` query. Let me go ahead and do that!

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:69]: Good point. I'll add that comment back in there 👍.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 4 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:85]: 69417e150ae7bbb6ad1f909fbb46e95bb39c1dd2 adds the above test.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:69]: 69417e150ae7bbb6ad1f909fbb46e95bb39c1dd2 re-adds this comment.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql [hunk 5]
```diff
@@ -18,157 +18,70 @@ import cpp
...
+   * This configuration intentionally conflates addresses of fields and their object, and pointer offsets
+   * with their base pointer as this allows us to detect cases where an object's address flows to a
+   * return statement via a field. For example:
+   *
+   * ```cpp
+   * struct S { int x, y };
+   * int* test() {
+   *   S s;
+   *   return &s.x; // BAD: &s.x is an address of a variable on the stack.
+   * }
```
Comment (jketema) [inline cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:68]: There's no closing triple quotes after snippet.

File: cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql (hunk 4 above)
Comment (jketema) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:85]: Is this something that needs wider testing? LGTM or at least DCA?

Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql:86]: I don't think this particular change warrants extended testing, but this PR as a whole definitely needs to be tested with DCA. I've already started a run which should hopefully be backlinked from this PR.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  ) {
+    not f.isVirtual() and
+    init.getEnclosingFunction() = f
+  }
+
+  /**
+   * Holds if `instr` is an argument (or argument indirection) to a call, and
+   * `succ` is the corresponding initialization instruction in the call target.
+   */
+  private predicate flowThroughCallable(Instruction argument, Instruction parameter) {
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/MustFlow.qll:149]: Good point. I would like to add flow out at some point, but I'm not exactly sure how to do that in a must-flow way.

**Comment by MathiasVP** [comment]
DCA looks good. Merging 🎉

**Pull request merged**